src/libraries/System.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/JSSynchronizationContext.cs

```diff
@@ -0,0 +1,143 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Threading;
+using System.Threading.Channels;
+using System.Runtime;
+using System.Runtime.InteropServices;
+using System.Runtime.CompilerServices;
+using System.Collections.Concurrent;
+using QueueType = System.Threading.Channels.Channel<System.Runtime.InteropServices.JavaScript.JSSynchronizationContext.WorkItem>;
+
+namespace System.Runtime.InteropServices.JavaScript {
+    internal sealed unsafe class JSSynchronizationContext : SynchronizationContext {
+        public readonly Thread MainThread;
+
+        internal readonly struct WorkItem {
+            public readonly SendOrPostCallback Callback;
+            public readonly object? Data;
+            public readonly ManualResetEventSlim? Signal;
+
+            public WorkItem (SendOrPostCallback callback, object? data, ManualResetEventSlim? signal) {
+                Callback = callback;
+                Data = data;
+                Signal = signal;
+            }
+        }
+
+        private static JSSynchronizationContext? MainThreadSynchronizationContext;
+        private readonly QueueType Queue;
+        private readonly Action _DataIsAvailable;
+
+        private JSSynchronizationContext (Thread mainThread)
+            : this (
+                mainThread,
+                Channel.CreateUnbounded<WorkItem>(
+                    new UnboundedChannelOptions { SingleWriter = false, SingleReader = true, AllowSynchronousContinuations = true }
+                )
+            )
+        {
+        }
+
+        private JSSynchronizationContext (Thread mainThread, QueueType queue) {
+            MainThread = mainThread;
+            Queue = queue;
+            _DataIsAvailable = DataIsAvailable;
+        }
+
+        public override SynchronizationContext CreateCopy () {
+            return new JSSynchronizationContext(MainThread, Queue);
+        }
+
+        private void AwaitNewData () {
+            var vt = Reader.WaitToReadAsync();
+            if (vt.IsCompleted) {
+                DataIsAvailable();
+                return;
+            }
+
+            var awaiter = vt.AsTask().ConfigureAwait(false).GetAwaiter();
+            awaiter.UnsafeOnCompleted(_DataIsAvailable);
+        }
+
+        private void DataIsAvailable () {
+            var onMainThread = Thread.CurrentThread == MainThread;
+            if (onMainThread)
+                Pump(Queue);
+            else
+                DispatchToMainThread(0, (void*)(delegate* unmanaged[Cdecl]<IntPtr, int, void>)&RequestPumpCallback, IntPtr.Zero, -1);
```

Comment:
`ScheduleBackgroundJob` (nee `mono_threads_schedule_background_job`) already knows to check if it's on the main thread or a worker.  So you don't need to dispatch calls to it to the main thread.

ie, I think you can just say

```csharp
   if (onMainThread)
        Pump(Queue);
   else
       ScheduleBackgroundJob((void*)(delegate* unmanaged[Cdecl]<void>)&BackgroundJobHandler);
```